### PR TITLE
Entities table&filters (#111 / #112)

### PR DESCRIPTION
**Pull request by Valecant**
## Proposed changes

Create a table where the Entities created by the user are displayed (#111) and can be filtered (#112)

## Types of changes

What types of changes does your code introduce to the project: _Put
an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation (improve documentation)
- [ ] Continuous Integration (functionality that improve automation or testing)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to
ask. We're here to help! This is simply a reminder of what we are going
to look for before merging your code._

- [x] I have read the [CONTRIBUTING][contrib] doc
- [x] I have signed the [Contributor License Agreement][cla]
- [x] I have updated the [RELEASE NOTES][release]
- [ ] I have added tests that prove my fix is effective or that my
      feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in
      downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion
by explaining why you chose the solution you did and what alternatives
you considered, etc...

[cla]: https://raw.githubusercontent.com/orchestracities/auth-management-ui/master/individual_cla.pdf
    "Martel Open Source Software Individual Contributor License Agreement"
[contrib]: https://github.com/orchestracities/auth-management-ui/blob/master/CONTRIBUTING.md
    "Contributing to auth-management-ui"
[release]: https://github.com/orchestracities/auth-management-ui/blob/master/RELEASE_NOTES.md
    "auth-management-ui Release Notes"

### CODE REVIEW

**Comment by chicco785** [comment]
1.  Always refer to the issue in the description (I already fixed it)
2.  Check relevant items in `Type of changes` and `Check list`

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -4,6 +4,8 @@
 
 ### New features
 
+- Entities Filter and Table display
```
Comment (chicco785) [inline RELEASE_NOTES.md:7]: - Table to list Entities (#111)
- Filter entities in table by servicePath, type and date (#112)

File: package.json [hunk 2]
```diff
@@ -7,22 +7,25 @@
     "@axa-fr/react-oidc": "^5.13.5",
     "@babel/eslint-parser": "^7.19.1",
     "@emotion/react": "latest",
-    "@emotion/styled": "v11.6.0",
-    "@mui/icons-material": "^5.6.2",
+    "@emotion/styled": "latest",
```
Comment (chicco785) [inline package.json:10]: never use latest, use always a fixed version

File: package.json [hunk 3]
```diff
@@ -7,22 +7,25 @@
     "@axa-fr/react-oidc": "^5.13.5",
     "@babel/eslint-parser": "^7.19.1",
     "@emotion/react": "latest",
```
Comment (chicco785) [inline package.json:9]: never use latest, use always a fixed version

File: package.json [hunk 4]
```diff
@@ -7,22 +7,25 @@
...
-    "@emotion/styled": "v11.6.0",
-    "@mui/icons-material": "^5.6.2",
+    "@emotion/styled": "latest",
+    "@mui/icons-material": "^5.10.15",
     "@mui/material": "latest",
     "@mui/system": "^5.10.15",
     "@mui/utils": "^5.10.15",
+    "@mui/x-date-pickers": "latest",
```
Comment (chicco785) [inline package.json:15]: never use latest, use always a fixed version

File: src/App.js [hunk 5]
```diff
@@ -536,6 +538,21 @@ export default class App extends Component {
                           </AuthorizedElement>
                         }
                       />
+                      <Route
+                        path="Entities"
+                        element={
+                          <AuthorizedElement tokenDecoded={this.state.tokenData} iSuperAdmin={true}>
```
Comment (chicco785) [inline src/App.js:544]: also not admin should be able to see entities (then which ones it's another matter, and it should be tackled by the API)

File: src/App.js [hunk 6]
```diff
@@ -536,6 +538,21 @@ export default class App extends Component {
...
+                            <EntitiesPage
+                              token={this.props.accessToken}
+                              tokenData={this.state.tokenData}
+                              env={env}
+                              graphqlErrors={this.state.connectionIssue}
```
Comment (chicco785) [inline src/App.js:549]: why graphql?

File: src/components/entities/entitiesTable.js [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+      <TableHead>
+        <TableRow>
+          <TableCell padding="checkbox">
+            <Checkbox
+              color="primary"
+              indeterminate={numSelected > 0 && numSelected < rowCount}
+              checked={rowCount > 0 && numSelected === rowCount}
+              onChange={onSelectAllClick}
+              inputProps={{
+                'aria-label': 'select all desserts'
```
Comment (chicco785) [inline src/components/entities/entitiesTable.js:167]: ?

File: src/components/entities/entitiesTable.js [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+                ) : null}
+              </TableSortLabel>
+            </TableCell>
+          ))}
+        </TableRow>
+      </TableHead>
+    );
+  };
+
+  PoliciesTableHead.propTypes = {
```
Comment (chicco785) [inline src/components/entities/entitiesTable.js:197]: ?

File: src/components/entities/entitiesTable.js [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+    numSelected: PropTypes.number.isRequired,
+    onRequestSort: PropTypes.func.isRequired,
+    onSelectAllClick: PropTypes.func.isRequired,
+    order: PropTypes.oneOf(['asc', 'desc']).isRequired,
+    orderBy: PropTypes.string.isRequired,
+    rowCount: PropTypes.number.isRequired
+  };
+
+  const PoliciesTableToolbar = (props) => {
```
Comment (chicco785) [inline src/components/entities/entitiesTable.js:206]: ?

File: src/components/entities/entitiesTable.js [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+          pl: { sm: 2 },
+          pr: { xs: 1, sm: 1 },
+          ...(numSelected > 0 && {
+            bgcolor: (theme) => alpha(theme.palette.primary.main, theme.palette.action.activatedOpacity)
+          })
+        }}
+      >
+        {numSelected > 0 ? (
+          <Typography sx={{ flex: '1 1 100%' }} color="inherit" variant="subtitle1" component="div">
+            <Trans i18nKey="policies.table.selected" values={{ name: numSelected }} />
```
Comment (chicco785) [inline src/components/entities/entitiesTable.js:221]: ?

File: src/components/entities/entitiesTable.js [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+
+        {numSelected > 0 ? (
+          <Tooltip title={<Trans>common.deleteTooltip</Trans>}>
+            <IconButton onClick={handleClickOpenDeleteDialog}>
+              <DeleteIcon />
+            </IconButton>
+          </Tooltip>
+        ) : (
+          <Trans
+            i18nKey="policies.table.total_plur"
```
Comment (chicco785) [inline src/components/entities/entitiesTable.js:235]: ?

File: src/components/entities/entitiesTable.js [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+            values={{
+              name: stableSort(rows, getComparator(order, orderBy)).length
+            }}
+          />
+        )}
+      </Toolbar>
+    );
+  };
+
+  PoliciesTableToolbar.propTypes = {
```
Comment (chicco785) [inline src/components/entities/entitiesTable.js:245]: again policies... ?

File: src/components/entities/entitiesTable.js [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+  // Avoid a layout jump when reaching the last page with empty rows.
+  const emptyRows = page > 0 ? Math.max(0, (1 + page) * rowsPerPage - rows.length) : 0;
+
+  const theme = useTheme();
+  const fullScreen = useMediaQuery(theme.breakpoints.down('sm'));
+  return (
+    <>
+      <Box sx={{ width: '100%' }}>
+        <DinamicPaper sx={{ width: '100%', mb: 2, overflow: 'hidden' }} elevation={1} square={false}>
+          <PoliciesTableToolbar numSelected={selected.length} />
```
Comment (chicco785) [inline src/components/entities/entitiesTable.js:307]: why policies?

File: src/components/entities/filters/pathFilter.js [hunk 14]
```diff
@@ -0,0 +1,124 @@
+import * as React from 'react';
```
Comment (chicco785) [inline src/components/entities/filters/pathFilter.js:1]: didn't we have already a pathFilter?

File: src/components/entities/filters/typeFilter.js [hunk 15]
```diff
@@ -0,0 +1,120 @@
+import * as React from 'react';
```
Comment (chicco785) [inline src/components/entities/filters/typeFilter.js:1]: how is this filter populated?

File: src/pages/entitiesPage.js [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+    const headers =
+      servicePath !== null
+        ? {
+            'fiware-Service': GeTenantData('name'),
+            'fiware-ServicePath':
+              servicePath.path[servicePath.path.length - 1] === '/' ? servicePath.path : servicePath.path
+          }
+        : { 'fiware-Service': GeTenantData('name') };
+    axios
+      .get(env.ORION.replace('&orderBy=id', ',') + 'dateCreated,dateModified,*' + queryParameters, {
```
Comment (chicco785) [inline src/pages/entitiesPage.js:81]: mmm let's discuss about this

File: src/components/entities/filters/typeFilter.js (hunk 15 above)
Comment (Valecant) [reply src/components/entities/filters/typeFilter.js:1]: > how is this filter populated?

Dynamically with the data retrieved without the cloned values

**Review by chicco785 — CHANGES_REQUESTED** [review]
see comments

**Comment by chicco785** [comment]
looking at ui review, the date is not displayed

File: src/pages/entitiesPage.js (hunk 16 above)
Comment (Valecant) [reply src/pages/entitiesPage.js:81]: > mmm let's discuss about this

i think that is ok for the retriving to have the url without parameters and then add the filtering on the policy form too

Comment (chicco785) [reply src/pages/entitiesPage.js:81]: 1. as in the case of the policies, the url may be different (and so ideally it should be picked from the resource type `entity` for the tenant
2. the url may or may not contain arbitrary query. so you should strip the all query if any.

File: src/components/entities/filters/typeFilter.js (hunk 15 above)
Comment (chicco785) [reply src/components/entities/filters/typeFilter.js:1]: so this is populated from the entities data.
ideally you should retrieve it from the api:
`/v2/types`

so if we had a resource type `entity type`, that could store the endpoint to retrieve the types.
the query returns a list as follows: 
```
[
  {
    "type": "Car",
    "attrs": {
      "speed": {
        "types": [
          "Number"
        ]
      },
      "fuel": {
        "types": [
          "gasoline",
          "diesel"
        ]
      },
      "temperature": {
        "types": [
          "urn:phenomenum:temperature"
        ]
      }
    },
    "count": 12
  },
  {
    "type": "Room",
    "attrs": {
      "pressure": {
        "types": [
          "Number"
        ]
      },
      "humidity": {
        "types": [
          "percentage"
        ]
      },
      "temperature": {
        "types": [
          "urn:phenomenum:temperature"
        ]
      }
    },
    "count": 7
  }
]
```
the name of the type is in the attribute `type`.

Comment (chicco785) [reply src/components/entities/filters/typeFilter.js:1]: it's also fine to not handle this in this pr. the point though is that this is they way to know all entity types (also the ones for which there is no entity at the time being)

**Comment by chicco785** [comment]
if a servicePath is not selected, you should anyhow pass as service path: '/#'

File: src/pages/entitiesPage.js (hunk 16 above)
Comment (chicco785) [reply src/pages/entitiesPage.js:81]: @Valecant i refactored the resource type creation.
see:
- https://github.com/orchestracities/auth-management-ui/blob/entitiesTable%26Filters/.env.development
- https://github.com/orchestracities/auth-management-ui/blob/entitiesTable%26Filters/configuration-api/main/mongo/populateDB.js

so:
1. if entity resource type is available, load the entity endpoint from there and strip any query configured there to add the query required here.
2. if not available, load it from env.Orion and append + '/v2/entities?'

File: package.json (hunk 4 above)
Comment (chicco785) [reply package.json:15]: @Valecant to explain you why: if you use latest, build are not consistent. so i build something today, it works, i build something tomorrow, if they changed something in the code, the build fails

File: src/components/entities/filters/dateFilter.js [hunk 17]
```diff
@@ -0,0 +1,137 @@
+import * as React from 'react';
```
Comment (chicco785) [inline src/components/entities/filters/dateFilter.js:1]: date filters may be used in other views, so move them to "shared"

File: src/components/entities/filters/pathFilter.js (hunk 14 above)
Comment (chicco785) [reply src/components/entities/filters/pathFilter.js:1]: i assume pathFilter can be a shared component

File: src/pages/entitiesPage.js [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+    const headers =
+      servicePath !== null
+        ? {
+            'fiware-Service': GeTenantData('name'),
+            'fiware-ServicePath':
+              servicePath.path[servicePath.path.length - 1] === '/' ? servicePath.path : servicePath.path
+          }
+        : { 'fiware-Service': GeTenantData('name') };
+    axios
+      .get(env.ORION + '/v2/entities?attrs=dateCreated,dateModified,*' + queryParameters, {
```
Comment (chicco785) [inline src/pages/entitiesPage.js:81]: @Valecant i refactored the resource type creation.
see:

https://github.com/orchestracities/auth-management-ui/blob/entitiesTable%26Filters/.env.development
https://github.com/orchestracities/auth-management-ui/blob/entitiesTable%26Filters/configuration-api/main/mongo/populateDB.js
so:

1. if entity resource type is available, load the entity endpoint from there and strip any query configured there to add the query required here.
2. if not available, load it from env.Orion and append + '/v2/entities?' (see how I modified above)

File: src/App.js (hunk 6 above)
Comment (Valecant) [reply src/App.js:549]: i'm

> why graphql?

i'm passing the state of the application to the page if a connection error needs to be displayed is going to be displayed

File: src/components/entities/filters/typeFilter.js [hunk 19]
```diff
@@ -47,7 +47,7 @@ const StyledMenu = styled((props) => (
   }
 }));
 
-export default function ResourceTypeFilter({ data, status, setstatus, filterValue }) {
+export default function TypeFilter({ data, status, setstatus, filterValue }) {
```
Comment (chicco785) [inline src/components/entities/filters/typeFilter.js:50]: ResourceType and Type of Entities are not the same thing!

Comment (chicco785) [reply src/components/entities/filters/typeFilter.js:50]: let's talk to understand what you try to achieve here :)

File: RELEASE_NOTES.md (hunk 1 above)
Comment (chicco785) [reply RELEASE_NOTES.md]: this is not resolved, the issues are not referred in the RELEASE_NOTES

File: src/components/entities/entitiesTable.js [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+  [theme.breakpoints.up('xl')]: {
+    width: '100%'
+  },
+  borderRadius: 10
+}));
+
+export default function EntitiesTable({ data }) {
+  // DELETE
+  const [openDeleteDialog, setOpenDeleteDialog] = React.useState(false);
+  console.log(openDeleteDialog);
```
Comment (chicco785) [inline src/components/entities/entitiesTable.js:64]: don't use console.log... didn't we start to use a proper logging library?

File: src/components/entities/entitiesTable.js [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+
+  const handleClickOpenDeleteDialog = () => {
+    setOpenDeleteDialog(true);
+  };
+
+  const handleCloseDeleteDialog = () => {
+    setOpenDeleteDialog(false);
+  };
+  console.log(handleCloseDeleteDialog);
```
Comment (chicco785) [inline src/components/entities/entitiesTable.js:73]: don't use console.log... didn't we start to use a proper logging library?

**Review by chicco785 — APPROVED** [review]
LGTM

**Pull request merged**